# <center>Reading Data from a CSV</center>

The purpose here is not to find a correlation between any features in this dataset. It's to understand how to manipulate the data so it's readable, easy to understand, usable to discover possible correlations between features, and make predictions based on the cleaned up data. That said, I thought while explaining this process, I'd use a dataset that is near and dear to my heart. Soccer is most definitely my favorite sport. I grew up playing and had an incredible coach that brought the best out in me, encouraged my natural competitive nature, taught me the skills I needed, and continued to fan the flame even as I went through high school. I am still in touch with him today. Everybody should have the opportunity to have such a great influence in their lives.

Moving on as that is not the point of this exercise. Have you ever wondered how the commentators at sporting events come up with specific stats on players as the game is being played? Yes, the obvious answer is someone behind the scenes feeds them the information. But, where does that person get the information from? People are not sitting there during a sporting event going through statistics by hand. Rarely if ever would a person find the right statistic for the right situation if having to manually find the relevant numbers.

Finding the numbers is also not enough. Multiple statistics are needed to make comparisons to past games, to know if an athlete has broken a world record or even broken their own record. How many times has a situation like the present one occurred? For instance, how many times has a penalty shot, during the regulation time while it was raining, gone in the net? That is a rather obscure statistic but believe it or not, commentators have presented crazier statistics than that example. 

So, where did all those statistics come from? Though we know it's someone that works with data, I'm referring to a more specific question. How does the person working with the data arrive at these statistics to give the commentators. That is what I will explain here. 

There are a few ways to accomplish that. Today, I will show you a basic way to be able to put the data in a readable format, clean it up, and group together features that might be connected.

Here, we will use Python to do the work we need. Specifically, we will use Pandas. In this case, Pandas is not referring to the adorable, fluffy, black, and white animals we've seen on TV or in zoos. Pandas is a package in Python designed for data science. Here I will show you the basics of what it can do and how it helps us access the data we're looking for. 

**The first thing we need to do is import the pandas library. When we import it, we are loading it into**
**memory so we can access it while we are working on this project. When we import it we shorten it by using pd.**
**Like other programming languages, Python has coding conventions.** 

In [2]:
# Here we import the pandas library as pd. 
import pandas as pd

Once the library is imported we do not need to import it again. We will have to reference it with pd.
When naming variables in Python, the convention is to use all lowercase letters.

The following line of code assigns the variable 'soccer' the value of the dataset.
In order to load the dataset, we refer back to our pandas library with pd and the function
'read_csv'. What does CSV stand for? It stands for 'Comma Separated Value' The url or the raw
dataset goes in quotes inside parentheses.

Another Python conventions is that variables be declared in lower case letters.

In [3]:
# The variable here is soccer. 
soccer = pd.read_csv('https://raw.githubusercontent.com/jalapic/engsoccerdata/master/data-raw/facup.csv')

# We want to see what the dataset looks like. To accomplish this, we can use the following code. Use the
# variable (here it is soccer) a dot, and then the word head with parentheses. 
soccer.head()

Date  Season                 home              visitor   FT  hgoal  \
0  1871-11-11    1871               Barnes        Civil Service  2-0    2.0   
1  1871-11-11    1871              Hitchin  Crystal Palace 1861  0-0    0.0   
2  1871-11-11    1871    Maidenhead United               Marlow  2-0    2.0   
3  1871-11-11    1871           Upton Park       Clapham Rovers  0-3    0.0   
4  1871-12-16    1871  Crystal Palace 1861    Maidenhead United  3-0    3.0   

   vgoal round      tie  aet  pen pens  hp  vp                   Venue  \
0    0.0     1  initial  NaN  NaN  NaN NaN NaN                     NaN   
1    0.0     1  initial  NaN  NaN  NaN NaN NaN      Top Field, Hitchin   
2    0.0     1  initial  NaN  NaN  NaN NaN NaN   York Road, Maidenhead   
3    3.0     1  initial  NaN  NaN  NaN NaN NaN   West Ham Park, London   
4    0.0     2  initial  NaN  NaN  NaN NaN NaN  Crystal Palace, London   

  attendance nonmatch                           notes neutral  
0      1,200      NaN                             NaN     NaN  
1        750      NaN  No replay, both teams progress     NaN  
2      1,287      NaN                             NaN     NaN  
3      1,500      NaN                             NaN     NaN  
4        NaN      NaN                             NaN     NaN

soccer.head() gives us the first 5 rows of the dataframe. That can give us a great amount of information.
If we want to see more rows at one time, we can just put a number inside the parentheses like so:

In [4]:
soccer.head(10) # The first 10 rows of the dataframe

Date  Season                 home              visitor   FT  hgoal  \
0  1871-11-11    1871               Barnes        Civil Service  2-0    2.0   
1  1871-11-11    1871              Hitchin  Crystal Palace 1861  0-0    0.0   
2  1871-11-11    1871    Maidenhead United               Marlow  2-0    2.0   
3  1871-11-11    1871           Upton Park       Clapham Rovers  0-3    0.0   
4  1871-12-16    1871  Crystal Palace 1861    Maidenhead United  3-0    3.0   
5  1871-12-16    1871            Wanderers       Clapham Rovers  3-1    3.0   
6  1871-12-23    1871               Barnes   Hampstead Heathens  1-1    1.0   
7  1872-01-06    1871   Hampstead Heathens               Barnes  1-0    1.0   
8  1872-01-10    1871              Hitchin      Royal Engineers  0-5    0.0   
9  1872-01-20    1871            Wanderers  Crystal Palace 1861  0-0    0.0   

   vgoal round      tie  aet  pen pens  hp  vp                   Venue  \
0    0.0     1  initial  NaN  NaN  NaN NaN NaN                     NaN   
1    0.0     1  initial  NaN  NaN  NaN NaN NaN      Top Field, Hitchin   
2    0.0     1  initial  NaN  NaN  NaN NaN NaN   York Road, Maidenhead   
3    3.0     1  initial  NaN  NaN  NaN NaN NaN   West Ham Park, London   
4    0.0     2  initial  NaN  NaN  NaN NaN NaN  Crystal Palace, London   
5    1.0     2  initial  NaN  NaN  NaN NaN NaN                     NaN   
6    1.0     2  initial  NaN  NaN  NaN NaN NaN          Barnes, London   
7    0.0     2   replay  NaN  NaN  NaN NaN NaN                     NaN   
8    5.0     2  initial  NaN  NaN  NaN NaN NaN                     NaN   
9    0.0     3  initial  NaN  NaN  NaN NaN NaN  Clapham Common, London   

  attendance nonmatch                           notes neutral  
0      1,200      NaN                             NaN     NaN  
1        750      NaN  No replay, both teams progress     NaN  
2      1,287      NaN                             NaN     NaN  
3      1,500      NaN                             NaN     NaN  
4        NaN      NaN                             NaN     NaN  
5      2,000      NaN                             NaN     NaN  
6        NaN      NaN                             NaN     NaN  
7        600      NaN                             NaN     NaN  
8        750      NaN                             NaN     NaN  
9        NaN      NaN  No replay, both teams progress     NaN

Notice the index of the dataframe starts at zero so though it goes to 4 the first time and 9 the second, there are 5 and 10 rows shown. We could also look at the last 5 rows of the dataframe by using tail instead of head.

In [5]:
soccer.tail() # The last 5 rows of the dataframe

Date  Season            home           visitor   FT  hgoal  \
16776  2016-11-15    2016   Coventry City         Morecambe  2-1    2.0   
16777  2016-11-15    2016  Fleetwood Town         Southport  4-1    4.0   
16778  2016-11-16    2016   Brackley Town        Gillingham  4-3    4.0   
16779  2016-11-17    2016  Dover Athletic  Cambridge United  2-4    2.0   
16780  2017-05-27    2016         Arsenal           Chelsea  2-1    2.0   

       vgoal round      tie  aet  pen pens  hp  vp                  Venue  \
16776    1.0     1   replay  NaN  NaN  NaN NaN NaN                    NaN   
16777    1.0     1   replay  yes  NaN  NaN NaN NaN                    NaN   
16778    3.0     1   replay  yes  NaN  NaN NaN NaN                    NaN   
16779    4.0     1   replay  yes  NaN  NaN NaN NaN                    NaN   
16780    1.0     f  initial  NaN  NaN  NaN NaN NaN  Wembley (new), London   

      attendance nonmatch notes neutral  
16776        NaN      NaN   NaN     NaN  
16777        NaN      NaN   NaN     NaN  
16778        NaN      NaN   NaN     NaN  
16779        NaN      NaN   NaN     NaN  
16780      89472      NaN   NaN     yes

We'll take a look at what information we can get from just what we see here but there is one more thing I'd like to 
know before picking this apart. I'd like to know how many rows and columns there are in this dataframe.

In [6]:
# We can do that with the following code
soccer.shape # The variable soccer, dot, and the word 'shape'. NO parentheses are used here.

(16781, 19)

The numbers shown in parentheses tell us there are 16,781 rows and 19 columns in this dataframe.

What do you think we should look at next? We want to make sure the data is readable and usable. Right now, I see many values that say *Nan*. We need to find out how many of those values exist is this data to be able to decide what to do with them. That can be done with the following code.

In [9]:
soccer.isnull() # This gives us everything in the dataframe using True for Nan, and false
# if it has a value. To make it easier to understand and less time consuming to evaluate, we
# can use sum() to get the value in each columns.

Date  Season   home  visitor     FT  hgoal  vgoal  round    tie  \
0      False   False  False    False  False  False  False  False  False   
1      False   False  False    False  False  False  False  False  False   
2      False   False  False    False  False  False  False  False  False   
3      False   False  False    False  False  False  False  False  False   
4      False   False  False    False  False  False  False  False  False   
5      False   False  False    False  False  False  False  False  False   
6      False   False  False    False  False  False  False  False  False   
7      False   False  False    False  False  False  False  False  False   
8      False   False  False    False  False  False  False  False  False   
9      False   False  False    False  False  False  False  False  False   
10     False   False  False    False  False  False  False  False  False   
11     False   False  False    False  False  False  False  False  False   
12     False   False  False    False  False  False  False  False  False   
13     False   False  False    False  False  False  False  False  False   
14     False   False  False    False  False  False  False  False  False   
15      True   False  False    False   True   True   True  False  False   
16      True   False  False    False   True   True   True  False  False   
17      True   False  False     True   True   True   True  False  False   
18      True   False  False    False   True   True   True  False  False   
19      True   False  False    False   True   True   True  False  False   
20      True   False  False     True   True   True   True  False  False   
21      True   False  False    False   True   True   True  False  False   
22     False   False  False    False  False  False  False  False  False   
23     False   False  False    False  False  False  False  False  False   
24     False   False  False    False  False  False  False  False  False   
25     False   False  False    False  False  False  False  False  False   
26     False   False  False    False  False  False  False  False  False   
27     False   False  False    False  False  False  False  False  False   
28     False   False  False    False  False  False  False  False  False   
29     False   False  False    False  False  False  False  False  False   
...      ...     ...    ...      ...    ...    ...    ...    ...    ...   
16751  False   False  False    False  False  False  False  False  False   
16752  False   False  False    False  False  False  False  False  False   
16753  False   False  False    False  False  False  False  False  False   
16754  False   False  False    False  False  False  False  False  False   
16755  False   False  False    False  False  False  False  False  False   
16756  False   False  False    False  False  False  False  False  False   
16757  False   False  False    False  False  False  False  False  False   
16758  False   False  False    False  False  False  False  False  False   
16759  False   False  False    False  False  False  False  False  False   
16760  False   False  False    False  False  False  False  False  False   
16761  False   False  False    False  False  False  False  False  False   
16762  False   False  False    False  False  False  False  False  False   
16763  False   False  False    False  False  False  False  False  False   
16764  False   False  False    False  False  False  False  False  False   
16765  False   False  False    False  False  False  False  False  False   
16766  False   False  False    False  False  False  False  False  False   
16767  False   False  False    False  False  False  False  False  False   
16768  False   False  False    False  False  False  False  False  False   
16769  False   False  False    False  False  False  False  False  False   
16770  False   False  False    False  False  False  False  False  False   
16771  False   False  False    False  False  False  False  False  False   
16772  False   False  False    Fa

In [10]:
soccer.isnull().sum()
# Here we're shown the column name on the left and the number of None Values on the right.

Date            282
Season            0
home              0
visitor         163
FT              284
hgoal           341
vgoal           341
round             0
tie               3
aet           16524
pen           16687
pens          16687
hp            16687
vp            16687
Venue         16271
attendance    15835
nonmatch      16440
notes         16437
neutral       16288
dtype: int64

There are a total of 16,781 rows. We see quite a few columns with a ton of None Values in them. That large number compared to the total number of rows probably make that data of little use. We can just get rid of those columns. There are few ways to do this. The safest way to do this would be to make a new dataframe. It leaves our soccer dataframe as is so we can reference it again if we need to. Here's how to make a copy using only the columns that are useful here.

In [19]:
useful_soccer = soccer[['Date', 'Season', 'home', 'visitor', 'FT', 'hgoal', 'vgoal', 'round', 'tie']]
# Let's see what that gives us by using the .head() command again
useful_soccer.head()

Date  Season                 home              visitor   FT  hgoal  \
0  1871-11-11    1871               Barnes        Civil Service  2-0    2.0   
1  1871-11-11    1871              Hitchin  Crystal Palace 1861  0-0    0.0   
2  1871-11-11    1871    Maidenhead United               Marlow  2-0    2.0   
3  1871-11-11    1871           Upton Park       Clapham Rovers  0-3    0.0   
4  1871-12-16    1871  Crystal Palace 1861    Maidenhead United  3-0    3.0   

   vgoal round      tie  
0    0.0     1  initial  
1    0.0     1  initial  
2    0.0     1  initial  
3    3.0     1  initial  
4    0.0     2  initial

Wow! That looks much easier to work with doesn't it? We still have None Values here though and we need to address those. Let's see where they are again. How do we find those? Rmember, we have a new variable for the dataframe we're using.

In [20]:
useful_soccer.tail()

Date  Season            home           visitor   FT  hgoal  \
16776  2016-11-15    2016   Coventry City         Morecambe  2-1    2.0   
16777  2016-11-15    2016  Fleetwood Town         Southport  4-1    4.0   
16778  2016-11-16    2016   Brackley Town        Gillingham  4-3    4.0   
16779  2016-11-17    2016  Dover Athletic  Cambridge United  2-4    2.0   
16780  2017-05-27    2016         Arsenal           Chelsea  2-1    2.0   

       vgoal round      tie  
16776    1.0     1   replay  
16777    1.0     1   replay  
16778    3.0     1   replay  
16779    4.0     1   replay  
16780    1.0     f  initial

In [17]:
useful_soccer.isnull().sum()

Date       282
Season       0
home         0
visitor    163
FT         284
hgoal      341
vgoal      341
round        0
tie          3
dtype: int64

The number of rows we have hasn't changed. Only the number of columns decreased when we chose not to use them. These None Values or null values are small compared to the total number of rows. Here we need to decide, do we just get rid of those rows? Do we fill them in with a different value? And, if so, what value do we choose? 

We see our first column to address is Date. Do you think the date matters? It could. We aren't sure exactly what we're trying to find. My reference earlier was to commentators using statistics during games. If we use the data for that purpose, the Date is very important. However, 282 rows out of 16,781 is minimal. As with all null values, we can get rid of them, or change them to something like the average, the mode, the median. That decision is individual and different based on the data you are working with. Let's take a quick look at the minimum, maximum, average, mode, median and quartile values in this data. When we do this, we will only get those values that have columns with numbers. We'll talk about the categorical columns next.

In [22]:
useful_soccer.describe()
# Here there are only three columns that return numbers to us. The Season's year, the number of goals 
# scored at a home game, and the number of goals scored at an away game. 

Season         hgoal         vgoal
count  16781.000000  16440.000000  16440.000000
mean    1957.485609      1.880657      1.230657
std       38.860727      1.705415      1.260164
min     1871.000000      0.000000      0.000000
25%     1930.000000      1.000000      0.000000
50%     1963.000000      2.000000      1.000000
75%     1989.000000      3.000000      2.000000
max     2016.000000     26.000000     14.000000

The season's year has no null values here. The number of goals scored at home and away has the same number of null values. Replacing those null values with the mean would not affect our data for those columns much at all so that would be a good value to use. For home goals the mean is 1.880657. For the away goals scored, the mean is 1.230657. We can use a command in pandas called .replace() to replace all the null values in each column with their means.

In [25]:
# There are several ways to accomplish this. There is another python library that makes this easy to
# do. It's call numpy and we'll import it like we did the pandas libary.
import numpy as np
useful_soccer['hgoal'] = useful_soccer['hgoal'].replace(np.nan, 1.880657)
# This says in the dataframe useful_soccer in the column hgoal, all null values will be filled
# with the mean and in fact, replace the null value.
useful_soccer.isnull().sum()

/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


Date       282
Season       0
home         0
visitor    163
FT         284
hgoal        0
vgoal      341
round        0
tie          3
dtype: int64